pkg/controllermanager/controller/shoot/shoot_hibernation_control.go

```diff
@@ -145,110 +57,292 @@ func (c *Controller) shootHibernationAdd(obj interface{}) {
 
 func (c *Controller) shootHibernationUpdate(oldObj, newObj interface{}) {
 	var (
-		oldShoot = oldObj.(*gardencorev1beta1.Shoot)
-		newShoot = newObj.(*gardencorev1beta1.Shoot)
-
-		oldSchedule = getShootHibernationSchedules(oldShoot)
-		newSchedule = getShootHibernationSchedules(newShoot)
+		oldShoot     = oldObj.(*gardencorev1beta1.Shoot)
+		newShoot     = newObj.(*gardencorev1beta1.Shoot)
+		oldFailed    = gutil.IsShootFailed(oldShoot)
+		newFailed    = gutil.IsShootFailed(newShoot)
+		oldSchedules = getShootHibernationSchedules(oldShoot.Spec.Hibernation)
+		newSchedules = getShootHibernationSchedules(newShoot.Spec.Hibernation)
 	)
 
-	if !reflect.DeepEqual(oldSchedule, newSchedule) {
-		key, err := cache.MetaNamespaceKeyFunc(newObj)
-		if err != nil {
-			gardenlogger.Logger.Errorf("Couldn't get key for object %+v: %v", newObj, err)
-			return
-		}
-		c.shootHibernationQueue.Add(key)
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
+	if err != nil {
+		gardenlogger.Logger.Errorf("Couldn't get key for object %+v: %v", newObj, err)
+		return
 	}
-}
 
-func (c *Controller) shootHibernationDelete(obj interface{}) {
-	shoot, ok := obj.(*gardencorev1beta1.Shoot)
-	if !ok {
+	if oldFailed && !newFailed && len(newSchedules) > 0 {
```

Comment:
As discussed via phone, let's not have this special handling for failed shoots in the event handler but just do nothing else than requeuing for the next event in the reconciler.